File: 226-invert-binary-tree/invert-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        # If the tree is empty, return None since there is nothing to invert
        if root is None:
            return 

        # Helper function to recursively swap the left and right children of each node
        def swap(node: Optional[TreeNode]) -> Optional[TreeNode]:
            # Base case: if the node is a leaf (no children), return the node as is
            if not node.left and not node.right:
                return node

            # Swap the left and right children of the current node
            tmpnode = node.left
            node.left = node.right
            node.right = tmpnode

            # Recursively apply the swap on the left and right subtrees
            # Check if there is a left child, and if so, recursively swap its children
            if node.left:
                node.left = swap(node.left)
            # Check if there is a right child, and if so, recursively swap its children
            if node.right:
                node.right = swap(node.right)

            # Return the current node after its children have been swapped
            return node

        # Call the swap function starting from the root node to invert the entire tree
        return swap(root)


        

File: 226-invert-binary-tree/invert-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode invertTree(TreeNode root) {
        if (root == null){
            return root;
        }
        TreeNode current = root;
        List<TreeNode> queue = new ArrayList<>();

        queue.add(current);

        while (!queue.isEmpty()){
            List<TreeNode> nxtLevel = new ArrayList<>();

            for (TreeNode node: queue){
                if (node.left != null){
                    nxtLevel.add(node.left);
                }
                if (node.right != null){
                    nxtLevel.add(node.right);
                }
                TreeNode left = node.left;
                TreeNode right = node.right;
                node.left = right;
                node.right = left;
            }
            queue = nxtLevel;
        }

        return root;
    }
}